From b2dbf9a61724f5d4c0a4d64944284a5a10fd2be5 Mon Sep 17 00:00:00 2001
From: David Vernet <void@manifault.com>
Date: Thu, 4 Jan 2024 16:19:13 -0600
Subject: [PATCH 063/152] scx: Error for a priq builtin DSQ in
 dispatch_enqueue()

We're currently checking whether a builtin DSQ is being used with priq
in scx_bpf_dispatch_vtime(). This neglects the fact that we could end up
falling back to scx_dsq_global if there's an error. If we error out with
SCX_ENQ_DSQ_PRIQ set in enqueue flags, we would trigger a warning in
dispatch_enqueue(). Let's instead just move the check to inside of
dispatch_enqueue().

Signed-off-by: David Vernet <void@manifault.com>
---
 kernel/sched/ext.c | 25 +++++++++++++------------
 1 file changed, 13 insertions(+), 12 deletions(-)

diff --git a/kernel/sched/ext.c b/kernel/sched/ext.c
index 7b4825ec1..9ff5f208d 100644
--- a/kernel/sched/ext.c
+++ b/kernel/sched/ext.c
@@ -646,8 +646,20 @@ static void dispatch_enqueue(struct scx_dispatch_q *dsq, struct task_struct *p,
 		}
 	}
 
+	if (unlikely((dsq->id & SCX_DSQ_FLAG_BUILTIN) &&
+		     (enq_flags & SCX_ENQ_DSQ_PRIQ))) {
+		/*
+		 * SCX_DSQ_LOCAL and SCX_DSQ_GLOBAL DSQs always consume from
+		 * their FIFO queues. To avoid confusion and accidentally
+		 * starving vtime-dispatched tasks by FIFO-dispatched tasks, we
+		 * disallow any internal DSQ from doing vtime ordering of
+		 * tasks.
+		 */
+		scx_ops_error("Cannot use vtime ordering for built-in DSQs");
+		enq_flags &= ~SCX_ENQ_DSQ_PRIQ;
+	}
+
 	if (enq_flags & SCX_ENQ_DSQ_PRIQ) {
-		WARN_ON_ONCE(dsq->id & SCX_DSQ_FLAG_BUILTIN);
 		p->scx.dsq_flags |= SCX_TASK_DSQ_ON_PRIQ;
 		rb_add_cached(&p->scx.dsq_node.priq, &dsq->priq,
 			      scx_dsq_priq_less);
@@ -4015,17 +4027,6 @@ void scx_bpf_dispatch_vtime(struct task_struct *p, u64 dsq_id, u64 slice,
 	if (!scx_dispatch_preamble(p, enq_flags))
 		return;
 
-	/*
-	 * SCX_DSQ_LOCAL and SCX_DSQ_GLOBAL DSQs always consume from their FIFO
-	 * queues. To avoid confusion and accidentally starving
-	 * vtime-dispatched tasks by FIFO-dispatched tasks, we disallow any
-	 * internal DSQ from doing vtime ordering of tasks.
-	 */
-	if (dsq_id & SCX_DSQ_FLAG_BUILTIN) {
-		scx_ops_error("Cannot use vtime ordering for built-in DSQs");
-		return;
-	}
-
 	if (slice)
 		p->scx.slice = slice;
 	else
-- 
2.43.0.232.ge79552d197

